iotdb/src/main/java/org/apache/iotdb/db/metadata/MManager.java

```diff
@@ -298,13 +363,105 @@ public void addPathToMTree(String path, String dataType, String encoding)
     TSDataType tsDataType = TSDataType.valueOf(dataType);
     TSEncoding tsEncoding = TSEncoding.valueOf(encoding);
     CompressionType type = CompressionType.valueOf(TSFileConfig.compressor);
-    addPathToMTree(path, tsDataType, tsEncoding, type, Collections.emptyMap());
+    addPathToMTreeInternal(path, tsDataType, tsEncoding, type, Collections.emptyMap());
+  }
+
+  private List<String> collectPaths(List<Path> paths) throws MetadataErrorException {
+    Set<String> pathSet = new HashSet<>();
+    // Attention: Monitor storage group seriesPath is not allowed to be deleted
+    for (Path p : paths) {
+      List<String> subPaths;
+      subPaths = getPaths(p.getFullPath());
+      if (subPaths.isEmpty()) {
+        throw new MetadataErrorException(String
+            .format("There are no timeseries in the prefix of %s seriesPath",
+                p.getFullPath()));
+      }
+      List<String> newSubPaths = new ArrayList<>();
+      for (String eachSubPath : subPaths) {
+        String filenodeName;
+        try {
+          filenodeName = getFileNameByPath(eachSubPath);
+        } catch (PathErrorException e) {
+          throw new MetadataErrorException(e);
+        }
+
+        if (MonitorConstants.STAT_STORAGE_GROUP_PREFIX.equals(filenodeName)) {
+          continue;
+        }
+        newSubPaths.add(eachSubPath);
+      }
+      pathSet.addAll(newSubPaths);
+    }
+    for (String p : pathSet) {
+      if (!pathExist(p)) {
+        throw new MetadataErrorException(String.format(
+            "Timeseries %s does not exist and cannot be deleted", p));
+      }
+    }
+    return new ArrayList<>(pathSet);
+  }
+
+  public Pair<Set<String>, Set<String>> deletePathFromMTree(String deletePath)
+      throws MetadataErrorException {
+    return deletePathFromMTree(new Path(deletePath));
+  }
+
+  public Pair<Set<String>, Set<String>> deletePathFromMTree(Path deletePath)
+      throws MetadataErrorException {
+    return deletePathsFromMTree(Collections.singletonList(deletePath));
+  }
+
+  public Pair<Set<String>, Set<String>> deletePathsFromMTree(List<Path> deletePathList)
```

Comment:
what is the meaning of the returned value?

Reply:
As the names of the returned variable suggest, StorageGroups to be closed and StorageGroups to be deleted.

Reply:
javadoc needed...

Reply:
added